CONTRIBUTING.md

```diff
@@ -456,6 +456,188 @@ checks if contributed code does not have the appropriate license headers.
 > be automatically configured to add the correct license header to new source
 > files based on the source location.
 
+### Logging
+
+The Elasticsearch server logs are vitally useful for diagnosing problems in a
+running cluster. You should make sure that your contribution uses logging
+appropriately: log enough detail to inform users about key events and help them
+understand what happened when things go wrong, without logging so much detail
+that the logs fill up with noise and the useful signal is lost.
+
+Elasticsearch uses Log4J for logging. In most cases you should log via a
+`Logger` named after the class that is writing the log messages, which you can
+do by declaring a static field of the class as follows:
+
+    private static final Logger logger = LogManager.getLogger(MyClassName.class);
```

Comment:
In 31cd3b8a10fc @nik9000 introduced a lot of new loggers with

```suggestion
    private static final Logger logger = LogManager.getLogger();
```

which has the advantage of being concise & copy-paste safe, but the disadvantage that it has to inspect the callstack to do so.

Can we decide which version we prefer?

Reply:
TIL, I didn't realise this did the right thing with subclasses, but it does. I don't have a strong opinion. What's the problem with inspecting the call stack?

Reply:
Huh. I think I don't particularly like that construct any more. It feels kind of wasteful for it to do a bunch of stuff with the stack to get the name of the logger to save you a couple of characters in the reference. I tend not to do it now. But I've never benchmarked it. If it's fast then I'm all for it.

I'll check the performance of the two. I think I wouldn't mind a little performance hit but but I don't want us to slow down start up to build loggers. I'll check!

Reply:
I benchmarked them and the no-arg version is about 38 times slower. About 1.7-ish
microseconds:

```
Benchmark                       Mode  Cnt     Score    Error  Units
LogManagerBenchmark.inferClass  avgt   10  1763.708 ± 14.058  ns/op
LogManagerBenchmark.passClass   avgt   10    46.341 ±  0.539  ns/op
```

I'm sure there's issues with it, but this is what I used:

```
@Fork(2)
@Warmup(iterations = 10)
@Measurement(iterations = 5)
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@State(Scope.Benchmark)
public class LogManagerBenchmark {
    @Benchmark
    public Object passClass() {
        return LogManager.getLogger(LogManagerBenchmark.class);
    }

    @Benchmark
    public Object inferClass() {
        return LogManager.getLogger();
    }
}
```


So, if we do the no-arg version of `getLogger` in every java file we'll spend
an extra 27 milliseconds on logger initialization when we load the classes.
That seems fine. As long as the logger is shared.